heartbeat/scheduler/scheduler.go

```diff
@@ -138,282 +178,102 @@ var ErrAlreadyStopped = errors.New("attempted to add job to already stopped sche
 
 // Add adds the given TaskFunc to the current scheduler. Will return an error if the scheduler
 // is done.
-func (s *Scheduler) Add(sched Schedule, id string, entrypoint TaskFunc) (removeFn func() error, err error) {
-	debugf("Add scheduler job '%v'.", id)
-
-	j := &job{
-		id:         id,
-		fn:         entrypoint,
-		schedule:   sched,
-		registered: false,
-		running:    0,
-	}
-	if s.isPreRunning() {
-		s.addSync(j)
-	} else if s.isRunning() {
-		s.addCh <- j
-	} else {
+func (s *Scheduler) Add(sched Schedule, id string, entrypoint TaskFunc) (removeFn context.CancelFunc, err error) {
+	if s.state.Load() == stateStopped {
 		return nil, ErrAlreadyStopped
 	}
 
-	return func() error { return s.remove(j) }, nil
-}
+	jobCtx, jobCtxCancel := context.WithCancel(s.ctx)
 
-func (s *Scheduler) remove(j *job) error {
-	debugf("Remove scheduler job '%v'", j.id)
+	// lastRanAt stores the last runAt the task was invoked
+	// The initial value is runAt.Now() because we use it to get the next runAt a job is scheduled to run
+	lastRanAt := time.Now().In(s.location)
 
-	if s.isPreRunning() {
-		s.doRemove(j)
-	} else if s.isRunning() {
-		s.rmCh <- j
-	}
-	// There is no need to handle the isDone case
-	// because removing the job accomplishes nothing if
-	// the scheduler is stopped
+	var taskFn timerqueue.TimerTaskFn
 
-	return nil
-}
+	schedNextRun := func() {
+		next := sched.Next(lastRanAt)
 
-func (s *Scheduler) run() {
-	defer func() {
-		// drain finished queue for active jobs to not leak
-		// go-routines on exit
-		for i := uint(0); i < s.active; i++ {
-			<-s.finished
+		now := time.Now().In(s.location)
+		if next.Before(now) {
+			// Our last invocation went long!
+			s.jobsMissedDeadline.Inc()
+			taskFn(now)
```

Comment:
is this an optimization to reuse the last go-routine? Can this impact a call to `Add`, which also calls `schedNextRun` as last step?

Reply:
This optimization is now removed anyway, it was unnecessary with how efficient the timer is now.